### PR TITLE
[Revalidation] Pause if the ingestion pipeline is unhealthy

### PR DESCRIPTION
**Pull request by loic-sharma**
This integrates with @scottbommarito's status work to pause the job if the ingestion pipeline is unhealthy.

Part of https://github.com/NuGet/Engineering/issues/1128

### CODE REVIEW

File: src/Validation.Common.Job/Validation.Common.Job.csproj [hunk 1]
```diff
@@ -115,7 +115,7 @@
       <Version>2.27.0</Version>
     </PackageReference>
     <PackageReference Include="NuGetGallery.Core">
-      <Version>4.4.5-dev-35743</Version>
+      <Version>4.4.5-loshar-health-36421</Version>
```
Comment (loic-sharma) [inline src/Validation.Common.Job/Validation.Common.Job.csproj]: I will update this before merging this change.

File: src/NuGet.Services.Revalidate/Services/HealthService.cs [hunk 2]
```diff
@@ -1,17 +1,52 @@
...
+                var status = JsonConvert.DeserializeObject<ServiceStatus>(json);
+                var component = status.ServiceRootComponent.GetByPath(_config.ComponentPath);
+
+                if (component == null)
+                {
+                    _logger.LogError(
+                        "Could not find component path {ComponentPath} in status health blob!",
+                        _config.ComponentPath);
+
+                    throw new InvalidOperationException($"Could not find component path '{_config.ComponentPath}' in status health blob!");
```
Comment (joelverhagen) [inline src/NuGet.Services.Revalidate/Services/HealthService.cs]: So if this exception is thrown, how will the caller treat it? Can we just return `false` in this case?

File: tests/NuGet.Services.Revalidate.Tests/Services/HealthServiceFacts.cs [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+        [Fact]
+        public async Task ThrowsIfStorageServiceThrows()
+        {
+            // This may happen if the status blob can't be found.
+            _storage
+                .Setup(s => s.GetFileAsync(_config.ContainerName, _config.StatusBlobName))
+                .ThrowsAsync(new Exception());
+
+            await Assert.ThrowsAsync<Exception>(() => _target.IsHealthyAsync());
```
Comment (joelverhagen) [inline tests/NuGet.Services.Revalidate.Tests/Services/HealthServiceFacts.cs]: Nit `Assert.Same` on the exception instance to verify no wrapping.

File: tests/NuGet.Services.Revalidate.Tests/Services/HealthServiceFacts.cs [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+namespace NuGet.Services.Revalidate.Tests.Services
+{
+    public class HealthServiceFacts
+    {
+        private readonly Mock<ICoreFileStorageService> _storage;
+        private readonly HealthConfiguration _config;
+        private readonly HealthService _target;
+
+        public HealthServiceFacts()
```
Comment (joelverhagen) [inline tests/NuGet.Services.Revalidate.Tests/Services/HealthServiceFacts.cs]: Test th missing component path?

Comment (loic-sharma) [reply tests/NuGet.Services.Revalidate.Tests/Services/HealthServiceFacts.cs]: Unless I'm misunderstanding you, this should be covered by [`ThrowsIfStatusBlobIsMissingExpectedPath`](https://github.com/NuGet/NuGet.Jobs/pull/504/files/c1f1fce8f291350a42e6ed11be65428d8b2e8762#diff-72997eca5a617c0c64f34c99485cf736R47).

File: src/NuGet.Services.Revalidate/Services/HealthService.cs (hunk 2 above)
Comment (loic-sharma) [reply src/NuGet.Services.Revalidate/Services/HealthService.cs]: If this exception is thrown, it'll be caught and reported by the `JobRunner`. The job would then exit and restart.

Returning `false` if the path can't be found sounds fine to me too.

File: tests/NuGet.Services.Revalidate.Tests/Services/HealthServiceFacts.cs (hunk 3 above)
Comment (loic-sharma) [reply tests/NuGet.Services.Revalidate.Tests/Services/HealthServiceFacts.cs]: Fixed

File: src/NuGet.Services.Revalidate/Services/HealthService.cs (hunk 2 above)
Comment (loic-sharma) [reply src/NuGet.Services.Revalidate/Services/HealthService.cs]: Updated

File: tests/NuGet.Services.Revalidate.Tests/Services/HealthServiceFacts.cs (hunk 4 above)
Comment (loic-sharma) [reply tests/NuGet.Services.Revalidate.Tests/Services/HealthServiceFacts.cs]: I renamed the test to [AssumesUnhealthyIfComponentCannotBeFoundInStatusBlob](https://github.com/NuGet/NuGet.Jobs/pull/504/files#diff-72997eca5a617c0c64f34c99485cf736R47)

Comment (joelverhagen) [reply tests/NuGet.Services.Revalidate.Tests/Services/HealthServiceFacts.cs]: Yup, perfect.

**Comment by loic-sharma** [comment]
Verified this on DEV with a manually created incident.

**Pull request merged**